Diabetes prediction by classification using structured vector machine
Workflow: 
1. Collect data 
2. data pre processing 
3. Train test split 
4. feed it to support vecotr machine classifier
5. Train Support vector machine classifier.


In [2]:
# libraries needed
import numpy as np #needed for numpy arrays
import pandas as pd #structuring needed
from sklearn.preprocessing import StandardScaler # to standarise data
from sklearn.model_selection import train_test_split # trauin test split
from sklearn import svm # supported vector machine
from sklearn.metrics import accuracy_score # to check accuracy of our model

Data collection and analysis

In [3]:
#PIMA Diabetes Datasetb

In [6]:
# loading the diabetes dataset
diabetes_dataset = pd.read_csv('diabetes.csv')

In [7]:
#printing the header
diabetes_dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [11]:
# number of rows and clolumns in dataset

diabetes_dataset.shape

(768, 9)

In [12]:
# 768 patients

In [13]:
#getting the statistical measures of the data
diabetes_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [14]:
diabetes_dataset['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

268 patients with diabetes, 500 no diabetes

In [17]:
diabetes_dataset.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

this shows us that patients mean. on average, diabetes patients have higher values for all columns.

In [20]:
#seperating data and labels
X = diabetes_dataset.drop(columns = 'Outcome', axis=1) # dropping the Outcome column
Y = diabetes_dataset['Outcome'] # setting Y to be the oucome column

In [21]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


we can see that Ouctome column is dropped

In [22]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


we have 767 rows of data 

Data standardization: Data fields have different ranges. we need to set a range of our standard data for our ML model to work.

In [26]:
scaler = StandardScaler()

In [27]:
scaler.fit(X)

StandardScaler()

In [28]:
standarized_data = scaler.transform(X)

In [29]:
print(standarized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


we can see that we have standarized our data

In [30]:
X = standarized_data # setting the standarized data to X.

print(X)
print(Y)

Train test split

In [32]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify=Y, random_state=2)
#20% of data is used for testing
#stratify to have equal cases of diabetic cases and non-diabetics cases in training data
#random state is fixed to 2 for consistent shuffling in this code.

In [35]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


Training model

In [37]:
# using suppported vector classifier
classifier = svm.SVC(kernel='linear')

In [75]:
#training the support vector classifier
classifier.fit(X_train, Y_train)

SVC(kernel='linear')

Model Evaluation

In [76]:
# accuracy score on training data
X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [77]:
print('Accuracy score of the training data : ', training_data_accuracy)

Accuracy score of the training data :  0.7866449511400652


In [78]:
# accuracy score on test data
X_test_prediction = classifier.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [79]:
print('Accuracy score of the training data : ', test_data_accuracy)

Accuracy score of the training data :  0.7727272727272727


Our model has accuracy score of ~77.2%

Making a predictive system

In [80]:
input_data = (1,163,72,0,0,39,1.222,33)

#changing the input data to numpy array
input_data_as_numpy_array = np.asarray(input_data)

#reshape the array as we are predicting one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

#standarizing the data
std_data = scaler.transform(input_data_reshaped)
print(std_data)

prediction = classifier.predict(std_data)
print(prediction)


if (prediction[0]==0):
    print('this person is not diabetic')
else:
    print('this person is  diabetic')

[[-0.84488505  1.31778097  0.14964075 -1.28821221 -0.69289057  0.8893767
   2.26546259 -0.0204964 ]]
[1]
this person is  diabetic


C:\Python\Python311\Lib\site-packages\sklearn\base.py:465: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
